### PR TITLE
REST API: Activate Related Posts module if not active when requesting related posts

### PR DESCRIPTION
**Pull request by tyxla**
It appears that currently if we request related posts through the non-cached endpoint, it'll fail in a bad way if module isn't enabled (a PHP fatal error). So this PR updates the endpoint to autoenable the module.

#### Changes proposed in this Pull Request:

* Enable Related Posts module if it's deactivated when requesting related posts through the endpoint.

#### Testing instructions:

* Follow test instructions of https://github.com/Automattic/wp-calypso/pull/27659
* Deactivate the related posts module.
* Insert a Related Posts block.
* Verify the module gets activated when the initial related posts request is made.

### CODE REVIEW

**Comment by jeherve** [comment]
I can't seem to get this to work today. I receive the following error in the editor when trying to insert the Related Posts block:

screenshot 2018-10-10 at 14 26 05

```
react-dom.min.82e21c65.js:110 TypeError: e.toUpperCase is not a function
    at r (main.min.js?ver=4.9.8:formatted:3878)
```

I also get PHP warnings from Gutenberg:

screenshot 2018-10-10 at 14 28 14

This seems to happen whether the Related Posts module is on or off, for some reason.

**Comment by tyxla** [comment]
@jeherve this might happen if you're using the development version of Gutenberg, and you've recently pulled the latest master, but forgot to run `npm run build` to build the new assets.

Could you try grabbing the latest master and rebuilding Gutenberg?

**Review by jeherve — DISMISSED** [review]
It works nicely now. 👍

**Comment by tyxla** [comment]
Thanks @jeherve! It would be great if anyone from @Automattic/luna could also review it before we 🚢

File: _inc/lib/core-api/class.jetpack-core-api-site-endpoints.php
Comment (simison) [inline _inc/lib/core-api/class.jetpack-core-api-site-endpoints.php]: I'm not very familiar with module activations, but seems like if we don't want to `redirect` or `exit` on module activation, it would be safer to do this here:

```php
Jetpack::activate_module( 'related-posts', false, false )
```

See https://github.com/Automattic/jetpack/blob/f29016b02aa81b1021283062e1e0943d2599bb70/class.jetpack.php#L2899

Comment (simison) [inline _inc/lib/core-api/class.jetpack-core-api-site-endpoints.php]: Nit: could these two `if`s be combined to one?

**Review by simison — DISMISSED** [review]
This indeed fixes the issue! 

Left a minor note about ensuring `activate_module()` behaving as expected regarding redirects.

File: _inc/lib/core-api/class.jetpack-core-api-site-endpoints.php
Comment (jeherve) [reply _inc/lib/core-api/class.jetpack-core-api-site-endpoints.php]: Maybe your question means we should make this statement more readable?

**Comment by kraftbj** [comment]
Marked as needs author reply pending the comment about the module activation maybe needing a comment to explain. This can go straight into `Ready to Merge` once that's been addressed either way.

File: _inc/lib/core-api/class.jetpack-core-api-site-endpoints.php
Comment (tyxla) [reply _inc/lib/core-api/class.jetpack-core-api-site-endpoints.php]: Good point 👍  I've done that in https://github.com/Automattic/jetpack/pull/10288/commits/7730aaa578ac392a8423c929d6a8e9eacf729cb4

Simply adding those was causing a fatal, though, because the related posts files module classes were not loaded when we activate a module after initialization. So we have to reload the modules after activating the module, which we're also doing in this commit.

**Comment by tyxla** [comment]
Feedback addressed, thanks folks!

I'd appreciate a final 👀 before shipping this - cc @jeherve @simison @kraftbj

**Comment by jeherve** [comment]
I got this to work and it seems to properly activate the module right now. 👍

**Comment by tyxla** [comment]
@kraftbj is there a reason for this one to be "In Progress"? The suggested solution already works well, and I've been waiting for a final review before 🚢

**Comment by kraftbj** [comment]
@tyxla I changed it since it sounded like a still active conversation (asking for suggestions, etc). If that conversation wasn't going to impact the PR, back to review.

**Review by jeherve — APPROVED** [review]
Approving since this worked for me.

**Comment by tyxla** [comment]
> why not returning an error or an empty result?

Well, we've been discussing what's best to be done in case module is deactivated, and it seems that returning an error when the module is inactive is a bad user experience and a point of confusion. Many of the authors/editors that are trying to insert a Related Posts block might not have an idea about the concept of modules, or if they have an idea about them, they wouldn't know how modules can be activated/deactivated. So we figured that if a user wishes to use related posts, they'll want the module to be active, thus we implicitly activate it.

> I am thinking about permissions, any user could activate the module?

This is a good point, but also brings us to the above answer. Just imagine how bad experience it would be if we error out here, for an author to try to insert the related posts block, and see that the module is deactivated, and then see that they don't have permissions to activate it. They'd have to ask the admin to activate it...not ideal IMHO, and definitely a bad user experience. 

I'd let @MichaelArestad chime in too in case I'm missing some points on UX in the case of deactivated related posts module.

**Comment by lezama** [comment]
Why would the user see the block if the module is deactivated?

**Comment by tyxla** [comment]
> Why would the user see the block if the module is deactivated?

Because we're trying to increase discoverability. We already have several conditions - Jetpack needs to be activated and connected for the blocks to appear there. If there are several more conditions that hide the block - the users might never find that this block even exists.

**Comment by MichaelArestad** [comment]
> Because we're trying to increase discoverability. We already have several conditions - Jetpack needs to be activated and connected for the blocks to appear there. If there are several more conditions that hide the block - the users might never find that this block even exists.

Let's shelve this idea for now. I think it's okay if they don't find the block if the module has been deactivated. Most folks will see it. Remember, module deactivation is different than the setting in Settings (hopefully).

**Comment by tyxla** [comment]
Alright, let's shut this down for now - I agree that a unified approach that address all other blocks that have a similar case would be a better idea 👍 

Thanks everyone! 🙌

**Pull request closed without merging**
